backstory/src/test/java/com/google/sps/servlets/AnalyzeImageServletTest.java

```diff
@@ -0,0 +1,178 @@
+    uploadedAnnotatedImages.add(uploadedAnnotatedImage);
+    when(mockImagesManager.createAnnotatedImagesFromImagesAsByteArrays(
+      any(List.class)
+    )).thenReturn(uploadedAnnotatedImages);
+
+    String sampleRawBackstory = "sampleRawBackstory";
+    when(mockStoryManager.generateText()).thenReturn(sampleRawBackstory);
+
+    StoryDecision sampleStoryDecision = new StoryDecision(sampleRawBackstory);
+    when(mockStoryAnalysisManager.generateDecision(anyString())).thenReturn(sampleStoryDecision);
+
+    // Create and set the factories to return the configured mocks.
+    BackstoryUserServiceFactory backstoryUserServiceFactory = () -> {
+      return mockUserService;
+    };
+    BackstoryDatastoreServiceFactory backstoryDatastoreServiceFactory = () -> {
+      return mockDatastoreService;
+    };
+    BlobstoreManagerFactory blobstoreManagerFactory = () -> {
+      return mockBlobstoreManager;
+    };
+    ImagesManagerFactory imagesManagerFactory = () -> {
+      return mockImagesManager;
+    };
+    StoryManagerFactory storyManagerFactory = (String prompt, int storyLength, double temperature) -> {
+      return mockStoryManager;
+    };
+    StoryAnalysisManagerFactory storyAnalysisManagerFactory = () -> {
+      return mockStoryAnalysisManager;
+    };
+    EntityFactory entityFactory = (String entityName) -> {
+      return mockAnalyzedImageEntity;
+    };
+    servlet.setBackstoryUserServiceFactory(backstoryUserServiceFactory);
+    servlet.setBlobstoreManagerFactory(blobstoreManagerFactory);
+    servlet.setBackstoryDatastoreServiceFactory(backstoryDatastoreServiceFactory);
+    servlet.setImagesManagerFactory(imagesManagerFactory);
+    servlet.setStoryManagerFactory(storyManagerFactory);
+    servlet.setStoryAnalysisManagerFactory(storyAnalysisManagerFactory);
+    servlet.setEntityFactory(entityFactory);
+
+    // doPost call to initiate testing.
+    servlet.doPost(mockRequest, mockResponse);
+
+    // Check that the request is as expected:
+    ArgumentCaptor<HttpServletRequest> requestCaptor = ArgumentCaptor.forClass(HttpServletRequest.class);
+    ArgumentCaptor<String> formInputNameCaptor = ArgumentCaptor.forClass(String.class);
+    verify(mockBlobstoreManager).getUploadedFileBlobKeyString(requestCaptor.capture(), formInputNameCaptor.capture());
+    // The formInputName must be "image-upload" because that is the name of the front-end form input element.
+    String expectedFormInputName = "image-upload";
+    String actualFormInputName = formInputNameCaptor.getValue();
+    Assert.assertEquals(expectedFormInputName, actualFormInputName);
+
+    // Verify that the proper properties are set for the entity
+    Text sampleBackstory = new Text(sampleRawBackstory);
+    verify(mockAnalyzedImageEntity).setProperty("userEmail", userEmail);
+    verify(mockAnalyzedImageEntity).setProperty("blobKeyString", blobKeyString);
+    verify(mockAnalyzedImageEntity).setProperty(eq("backstory"), any(Text.class));
```

Comment:
Are there any validations that can be made on this text? (we want to make sure the servlet is storing the right thing)

Reply:
Yeah, I could design some validations. The reason I can't just make sure the text inserted is equal the test backstory is because the story endings are randomly added. Do you have any ideas for possible validations?